### PR TITLE
Support multiple record batches, closes #1022

### LINKED ISSUE #1022: Single message consumer with 0.11.0 API
##### Versions

Sarama Version: f0c32558d5e8d914b3de1a6f504e2ca88aaa40c4
Kafka Version: 0.11.0.1
Go Version: go1.9.2

##### Configuration

Sarama is using Kafka API 0.11.0.0.

##### Logs

I added some debug logging, here's the output:

```
018/01/12 06:17:11 Request: max wait time = 500, min bytes = 1, max bytes = 104857600, isolation = 0, num blocks = 1
2018/01/12 06:17:11 Claimed logs.www.nginx.access: [1]
2018/01/12 06:17:11   fetch request block for partition 1: &sarama.fetchRequestBlock{fetchOffset:15168197722, maxBytes:10000}
2018/01/12 06:17:11 Request: max wait time = 500, min bytes = 1, max bytes = 104857600, isolation = 0, num blocks = 1
2018/01/12 06:17:11   fetch request block for partition 1: &sarama.fetchRequestBlock{fetchOffset:15168197722, maxBytes:10000}
2018/01/12 06:17:11 Got response: version = 4, numTopics = 1, size = 10057 bytes
2018/01/12 06:17:11   Num blocks: 1
2018/01/12 06:17:11   Remaining: 10030
2018/01/12 06:17:11    block recordsSize = 10000
2018/01/12 06:17:11    block recordsSize after = 10000
2018/01/12 06:17:11      decoding record type = 2
2018/01/12 06:17:11      decoding record batch: offset = 15168197722, delta = 0, len = 901
2018/01/12 06:17:11      decoding record batch: numRecs = 1 (9939 bytes remaining)
2018/01/12 06:17:11      decoding record batch: codec = 0 (9087 bytes remaining)
2018/01/12 06:17:11   Remaining after block 0 (1 / 1 records): 0
```

Here's the diff that added the logging:

```diff
diff --git a/vendor/github.com/Shopify/sarama/fetch_request.go b/vendor/github.com/Shopify/sarama/fetch_request.go
index 8c8e3a5..9701402 100644
--- a/vendor/github.com/Shopify/sarama/fetch_request.go
+++ b/vendor/github.com/Shopify/sarama/fetch_request.go
@@ -1,5 +1,7 @@
 package sarama
 
+import "log"
+
 type fetchRequestBlock struct {
 	fetchOffset int64
 	maxBytes    int32
@@ -54,6 +56,9 @@ func (r *FetchRequest) encode(pe packetEncoder) (err error) {
 	if err != nil {
 		return err
 	}
+
+	log.Printf("Request: max wait time = %d, min bytes = %d, max bytes = %d, isolation = %d, num blocks = %d", r.MaxWaitTime, r.MinBytes, r.MaxBytes, r.Isolation, len(r.blocks))
+
 	for topic, blocks := range r.blocks {
 		err = pe.putString(topic)
 		if err != nil {
@@ -65,12 +70,14 @@ func (r *FetchRequest) encode(pe packetEncoder) (err error) {
 		}
 		for partition, block := range blocks {
 			pe.putInt32(partition)
+			log.Printf("  fetch request block for partition %d: %#v", partition, block)
 			err = block.encode(pe)
 			if err != nil {
 				return err
 			}
 		}
 	}
+
 	return nil
 }
 
diff --git a/vendor/github.com/Shopify/sarama/fetch_response.go b/vendor/github.com/Shopify/sarama/fetch_response.go
index 86fa549..bb46305 100644
--- a/vendor/github.com/Shopify/sarama/fetch_response.go
+++ b/vendor/github.com/Shopify/sarama/fetch_response.go
@@ -1,6 +1,10 @@
 package sarama
 
-import "time"
+import (
+	"log"
+	"os"
+	"time"
+)
 
 type AbortedTransaction struct {
 	ProducerID  int64
@@ -75,10 +79,15 @@ func (b *FetchResponseBlock) decode(pd packetDecoder, version int16) (err error)
 		return err
 	}
 
+	log.Printf("   block recordsSize = %d", recordsSize)
+
 	recordsDecoder, err := pd.getSubset(int(recordsSize))
 	if err != nil {
 		return err
 	}
+
+	log.Printf("   block recordsSize after = %d", recordsSize)
+
 	if recordsSize > 0 {
 		if err = b.Records.decode(recordsDecoder); err != nil {
 			return err
@@ -136,6 +145,8 @@ func (r *FetchResponse) decode(pd packetDecoder, version int16) (err error) {
 		return err
 	}
 
+	log.Printf("Got response: version = %d, numTopics = %d, size = %d bytes", r.Version, numTopics, pd.remaining())
+
 	r.Blocks = make(map[string]map[int32]*FetchResponseBlock, numTopics)
 	for i := 0; i < numTopics; i++ {
 		name, err := pd.getString()
@@ -148,6 +159,9 @@ func (r *FetchResponse) decode(pd packetDecoder, version int16) (err error) {
 			return err
 		}
 
+		log.Printf("  Num blocks: %d", numBlocks)
+		log.Printf("  Remaining: %d", pd.remaining())
+
 		r.Blocks[name] = make(map[int32]*FetchResponseBlock, numBlocks)
 
 		for j := 0; j < numBlocks; j++ {
@@ -162,9 +176,14 @@ func (r *FetchResponse) decode(pd packetDecoder, version int16) (err error) {
 				return err
 			}
 			r.Blocks[name][id] = block
+
+			nr, _ := block.Records.numRecords()
+			log.Printf("  Remaining after block %d (%d / %d records): %d", j, len(block.Records.recordBatch.Records), nr, pd.remaining())
 		}
 	}
 
+	os.Exit(42)
+
 	return nil
 }
 
diff --git a/vendor/github.com/Shopify/sarama/record_batch.go b/vendor/github.com/Shopify/sarama/record_batch.go
index a8c533b..c798653 100644
--- a/vendor/github.com/Shopify/sarama/record_batch.go
+++ b/vendor/github.com/Shopify/sarama/record_batch.go
@@ -5,6 +5,7 @@ import (
 	"compress/gzip"
 	"fmt"
 	"io/ioutil"
+	"log"
 	"time"
 
 	"github.com/eapache/go-xerial-snappy"
@@ -123,6 +124,7 @@ func (b *RecordBatch) decode(pd packetDecoder) (err error) {
 	if err != nil {
 		return err
 	}
+
 	b.Codec = CompressionCodec(int8(attributes) & compressionCodecMask)
 	b.Control = attributes&controlMask == controlMask
 
@@ -130,6 +132,8 @@ func (b *RecordBatch) decode(pd packetDecoder) (err error) {
 		return err
 	}
 
+	log.Printf("     decoding record batch: offset = %d, delta = %d, len = %d", b.FirstOffset, b.LastOffsetDelta, batchLen)
+
 	if err = (Timestamp{&b.FirstTimestamp}).decode(pd); err != nil {
 		return err
 	}
@@ -154,6 +158,9 @@ func (b *RecordBatch) decode(pd packetDecoder) (err error) {
 	if err != nil {
 		return err
 	}
+
+	log.Printf("     decoding record batch: numRecs = %d (%d bytes remaining)", numRecs, pd.remaining())
+
 	if numRecs >= 0 {
 		b.Records = make([]*Record, numRecs)
 	}
@@ -173,6 +180,8 @@ func (b *RecordBatch) decode(pd packetDecoder) (err error) {
 		return err
 	}
 
+	log.Printf("     decoding record batch: codec = %d (%d bytes remaining)", b.Codec, pd.remaining())
+
 	switch b.Codec {
 	case CompressionNone:
 	case CompressionGZIP:
diff --git a/vendor/github.com/Shopify/sarama/records.go b/vendor/github.com/Shopify/sarama/records.go
index 54ee7e3..ec873f5 100644
--- a/vendor/github.com/Shopify/sarama/records.go
+++ b/vendor/github.com/Shopify/sarama/records.go
@@ -1,6 +1,9 @@
 package sarama
 
-import "fmt"
+import (
+	"fmt"
+	"log"
+)
 
 const (
 	unknownRecords = iota
@@ -90,6 +93,8 @@ func (r *Records) decode(pd packetDecoder) error {
 		}
 	}
 
+	log.Printf("     decoding record type = %d", r.recordsType)
+
 	switch r.recordsType {
 	case legacyRecords:
 		r.msgSet = &MessageSet{}

```

##### Problem Description

This is related to #901.

See the ticket I filed previously for Kafka itself: https://issues.apache.org/jira/browse/KAFKA-6441

The problem is that Sarama decodes 0.11.0.0 `FetchResponse` in such a way, that only the very first message is extracted. We have pretty deep buffers and this resulted in Kafka outbound jumping from 125Mbit/s to 55000Mbit/s for a tiny topic. Essentially, consumers are only able to advance 1 message at a time, while Kafka always replies with filled buffers with thousands of messages.

I'm not sure if the problem is within Kafka or Sarama.

In reply for the logged request above we get this reply (truncated):

```go
	reply := []byte{
		// base offset: 8 bytes
		0x0, 0x0, 0x0, 0x3, 0x88, 0x18, 0x54, 0x5a,
		// len: 4 bytes
		0x0, 0x0, 0x3, 0x85,
		// leader epoch: 4 bytes,
		0x0, 0x0, 0x0, 0x8c,
		// version: 1 byte,
		0x2,
		// crc: 4 bytes
		0x78, 0xb4, 0x40, 0x58,
		// attributes: 2 bytes
		0x0, 0x0,
		// last offset delta: 4 bytes
		0x0, 0x0, 0x0, 0x0,
		// first timestamp: 8 bytes
		0x0, 0x0, 0x1, 0x60, 0xe9, 0x3, 0x46, 0x1e,
		// last timestamp: 8 bytes
		0x0, 0x0, 0x1, 0x60, 0xe9, 0x3, 0x46, 0x1e,
		// producer id: 8 bytes
		0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
		// producer epoch: 2 bytes
		0xff, 0xff,
		// base sequence: 4 bytes
		0xff, 0xff, 0xff, 0xff,
		// records count: 4 bytes
		0x0, 0x0, 0x0, 0x1,
		// except from the first record below
		0xa4, 0xd, 0x0, 0x0, 0x0, 0x10, 0x31, 0x30, 0x38, 0x63, 0x6f, 0x6d, 0x32,
		0x32, 0x86, 0xd, 0x7b, 0x22, 0x40, 0x74, 0x69, 0x6d, 0x65, 0x73, 0x74,
		0x61, 0x6d, 0x70, 0x22, 0x3a, 0x20, 0x22, 0x32, 0x30, 0x31, 0x38, 0x2d,
		0x30, 0x31, 0x2d, 0x31, 0x32, 0x54, 0x30, 0x36, 0x3a, 0x31, 0x36, 0x3a,
		0x33, 0x34, 0x2b, 0x30, 0x30, 0x3a, 0x30, 0x30, 0x22, 0x2c, 0x20, 0x22,
		0x40, 0x76, 0x65, 0x72, 0x73, 0x69, 0x6f, 0x6e, 0x22, 0x3a, 0x20, 0x31,
		0x2c, 0x22, 0x68, 0x6f, 0x73, 0x74, 0x22, 0x3a,
		// ... lots more bytes than just one record ...
	}
```

Kafka sets `recordsSize` in `FetchResponse` to a large number (`10000` in the logs), that is supposed to hold multiple messages (which are a lot shorter than `10000` bytes). Then in `RecordBatch` Kafka clearly says that there is only one record (`0x0, 0x0, 0x0, 0x1`) in this record and the length of the whole thing is 901 bytes (`0x3, 0x85`).

The gap between 901 bytes and 10000 bytes filled with other messages (records? record batches?), which Sarama does not see.

I looked at Kafka protocol docs, but they do not mention anything about this:

* https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol

cc @wladh

### PR DESCRIPTION
**Pull request by bobrik**
cc @wladh

### CODE REVIEW

**Comment by bobrik** [comment]
With added logging:

```diff
diff --git a/consumer.go b/consumer.go
index 276bc94..bb850f3 100644
--- a/consumer.go
+++ b/consumer.go
@@ -3,6 +3,7 @@ package sarama
 import (
        "errors"
        "fmt"
+       "log"
        "sync"
        "sync/atomic"
        "time"
@@ -619,6 +620,8 @@ func (child *partitionConsumer) parseResponse(response *FetchResponse) ([]*Consu
                }
        }

+       log.Printf("Returning %d messages from %d batches", len(messages), len(block.Records.recordBatchSet.batches))
+
        return messages, nil
 }
```

I see multiple batches on the consumer as expected (1MB buffer):

```
2018/01/14 03:34:23 Returning 2144 messages from 2054 batches
2018/01/14 03:34:23 Returning 2136 messages from 2033 batches
2018/01/14 03:34:23 Returning 2136 messages from 2005 batches
2018/01/14 03:34:23 Returning 2136 messages from 1973 batches
2018/01/14 03:34:23 Returning 2134 messages from 2017 batches
2018/01/14 03:34:23 Returning 2134 messages from 2035 batches
2018/01/14 03:34:23 Returning 2137 messages from 2012 batches
2018/01/14 03:34:23 Returning 2134 messages from 2005 batches
2018/01/14 03:34:23 Returning 2159 messages from 1993 batches
2018/01/14 03:34:23 Returning 2139 messages from 1948 batches
2018/01/14 03:34:23 Returning 2144 messages from 2018 batches
2018/01/14 03:34:23 Returning 2141 messages from 2033 batches
2018/01/14 03:34:23 Returning 1911 messages from 1781 batches
2018/01/14 03:34:23 Returning 54 messages from 53 batches
2018/01/14 03:34:23 Returning 13 messages from 13 batches
2018/01/14 03:34:23 Returning 28 messages from 28 batches
2018/01/14 03:34:24 Returning 7 messages from 7 batches
2018/01/14 03:34:24 Returning 50 messages from 48 batches
2018/01/14 03:34:24 Returning 4 messages from 3 batches
```

This producer is not great at batching, so numbers match here. You can see big number of batches after restart and them smaller batches as consumer is up to date with production rate.

On another consumer with higher throughput and better batching we see this:

```
2018/01/14 03:38:12 Returning 12673 messages from 165 batches
2018/01/14 03:38:12 Returning 13256 messages from 168 batches
2018/01/14 03:38:12 Returning 12839 messages from 169 batches
2018/01/14 03:38:12 Returning 12519 messages from 162 batches
2018/01/14 03:38:12 Returning 12783 messages from 166 batches
2018/01/14 03:38:12 Returning 12999 messages from 167 batches
2018/01/14 03:38:13 Returning 12856 messages from 163 batches
2018/01/14 03:38:13 Returning 12764 messages from 165 batches
2018/01/14 03:38:13 Returning 12873 messages from 165 batches
2018/01/14 03:38:13 Returning 13179 messages from 166 batches
2018/01/14 03:38:13 Returning 13196 messages from 168 batches
2018/01/14 03:38:13 Returning 10121 messages from 130 batches
2018/01/14 03:38:13 Returning 312 messages from 4 batches
2018/01/14 03:38:13 Returning 253 messages from 3 batches
2018/01/14 03:38:13 Returning 149 messages from 2 batches
2018/01/14 03:38:13 Returning 88 messages from 1 batches
2018/01/14 03:38:13 Returning 236 messages from 3 batches
2018/01/14 03:38:13 Returning 61 messages from 1 batches
2018/01/14 03:38:13 Returning 85 messages from 1 batches
```

**Comment by wladh** [comment]
While this fix addresses the issue of multiple record batches being in one response, there could be a mix of record batches and legacy message sets (see issue #1021). Fixing that issue needs to rework how records are implemented, because they aren't just either legacy or new anymore, and the ordering between record batches and legacy messages needs to be preserved.

**Comment by bobrik** [comment]
That's a great point. I updated the PR to support mixed scenarios.

Tested with 0.10.2.0 and 0.11.0.0 API with the following log:

```
# /usr/local/kafka/bin/kafka-run-class.sh kafka.tools.DumpLogSegments --files /state/kafka/dev/sarama.test-0/00000000000000000000.log
Dumping /state/kafka/dev/sarama.test-0/00000000000000000000.log
Starting offset: 0
offset: 0 position: 0 NoTimestampType: -1 isvalid: true size: 40 magic: 0 compresscodec: NONE crc: 396773912
offset: 1 position: 40 NoTimestampType: -1 isvalid: true size: 40 magic: 0 compresscodec: NONE crc: 396773912
baseOffset: 2 lastOffset: 2 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 80 CreateTime: 1516069549010 isvalid: true size: 82 magic: 2 compresscodec: NONE crc: 3770695923
baseOffset: 3 lastOffset: 3 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 162 CreateTime: 1516069549037 isvalid: true size: 82 magic: 2 compresscodec: NONE crc: 431266329
baseOffset: 4 lastOffset: 4 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 244 CreateTime: 1516074820185 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 2860755674
baseOffset: 5 lastOffset: 5 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 328 CreateTime: 1516074820185 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 2657354307
baseOffset: 7 lastOffset: 7 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 412 CreateTime: 1516074820198 isvalid: true size: 107 magic: 2 compresscodec: NONE crc: 2824390372
offset: 8 position: 519 NoTimestampType: -1 isvalid: true size: 42 magic: 0 compresscodec: NONE crc: 2108329865
offset: 9 position: 561 NoTimestampType: -1 isvalid: true size: 42 magic: 0 compresscodec: NONE crc: 3835944499
offset: 10 position: 603 NoTimestampType: -1 isvalid: true size: 42 magic: 0 compresscodec: NONE crc: 2108329865
offset: 11 position: 645 NoTimestampType: -1 isvalid: true size: 42 magic: 0 compresscodec: NONE crc: 3835944499
baseOffset: 12 lastOffset: 12 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 687 CreateTime: 1516074874413 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 795684345
baseOffset: 13 lastOffset: 13 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 771 CreateTime: 1516074874413 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 462062944
baseOffset: 14 lastOffset: 14 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 855 CreateTime: 1516074874420 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 1948247459
baseOffset: 15 lastOffset: 15 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 939 CreateTime: 1516074874420 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 1090014522
baseOffset: 16 lastOffset: 17 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false position: 1023 CreateTime: 1516077773608 isvalid: true size: 107 magic: 2 compresscodec: NONE crc: 293091933
baseOffset: 18 lastOffset: 18 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 1130 CreateTime: 1516077773613 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 696319396
baseOffset: 19 lastOffset: 19 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: false position: 1214 CreateTime: 1516077773613 isvalid: true size: 84 magic: 2 compresscodec: NONE crc: 493311293
```

I've tried both full (1KB covers everything in this log) and short (128B) reads, both work as expected.

File: fetch_response.go [hunk 1]
```diff
@@ -79,15 +82,51 @@ func (b *FetchResponseBlock) decode(pd packetDecoder, version int16) (err error)
 	if err != nil {
 		return err
 	}
-	if recordsSize > 0 {
-		if err = b.Records.decode(recordsDecoder); err != nil {
+
+	b.RecordsSet = []*Records{}
+
+	for {
+		if recordsDecoder.remaining() == 0 {
```
Comment (wladh) [inline fetch_response.go]: I think `for recordsDecoder.remaining() > 0 {` would be more idiomatic

File: record_batch.go [hunk 2]
```diff
@@ -35,6 +35,47 @@ func (e recordsArray) decode(pd packetDecoder) error {
 	return nil
 }
 
+type RecordBatchSet struct {
```
Comment (wladh) [inline record_batch.go]: we don't seem to use this type anywhere

File: records.go [hunk 3]
```diff
@@ -1,167 +1,72 @@
...
-	legacyRecords
-	defaultRecords
-
 	magicOffset = 16
 	magicLength = 1
 )
 
 // Records implements a union type containing either a RecordBatch or a legacy MessageSet.
 type Records struct {
-	recordsType int
```
Comment (wladh) [inline records.go]: The reason why I put this field is to make it easier to debug possible inconsistencies between what the Records should contain and what it actually contains. If we want to refactor this, we could do it in a separate PR because I don't think it's related to the problem at hand (unless I'm missing something).

File: fetch_response.go (hunk 1 above)
Comment (bobrik) [reply fetch_response.go]: That would require one more level of nesting, I think it's better to break early and keep it flat:

* https://softwareengineering.stackexchange.com/a/18459

File: records.go (hunk 3 above)
Comment (bobrik) [reply records.go]: Fair enough. I built on top of my initial changes, that's why many things were missing. I added another commit to minimize the diff with master.

File: fetch_response.go [hunk 4]
```diff
@@ -31,7 +33,8 @@ type FetchResponseBlock struct {
 	HighWaterMarkOffset int64
 	LastStableOffset    int64
 	AbortedTransactions []*AbortedTransaction
-	Records             Records
```
Comment (eapache) [inline fetch_response.go]: Technically it's a breaking change to remove this; could you leave it with a comment that it's deprecated, and just fill it in with the first set or something?

File: fetch_response.go (hunk 1 above)
Comment (eapache) [reply fetch_response.go]: I'm not sure I follow; moving the condition into the `for` loop that already exists wouldn't add any layers of nesting?

**Comment by eapache** [comment]
This looks pretty good to me, just a few comments. Thanks for all your work on this both of you!

File: fetch_response.go (hunk 1 above)
Comment (bobrik) [reply fetch_response.go]: My bad, I misread the first time. Code is changed now.

File: fetch_response.go (hunk 4 above)
Comment (bobrik) [reply fetch_response.go]: Done, let me know if that's what you had in mind.

**Comment by eapache** [comment]
Thanks!
